In [113]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns
import warnings 
warnings.filterwarnings("ignore") 

In [114]:
data = pd.read_csv("C:/Users/PC/Documents/archive-Ridwan/Airbnb_Open_Data.csv")
data.head()

id                                              NAME      host id  \
0  1001254                Clean & quiet apt home by the park  80014485718   
1  1002102                             Skylit Midtown Castle  52335172823   
2  1002403               THE VILLAGE OF HARLEM....NEW YORK !  78829239556   
3  1002755                                               NaN  85098326012   
4  1003689  Entire Apt: Spacious Studio/Loft by central park  92037596077   

  host_identity_verified host name neighbourhood group neighbourhood  \
0            unconfirmed  Madaline            Brooklyn    Kensington   
1               verified     Jenna           Manhattan       Midtown   
2                    NaN     Elise           Manhattan        Harlem   
3            unconfirmed     Garry            Brooklyn  Clinton Hill   
4               verified    Lyndon           Manhattan   East Harlem   

        lat      long        country  ... service fee minimum nights  \
0  40.64749 -73.97237  United States  ...       $193            10.0   
1  40.75362 -73.98377  United States  ...        $28            30.0   
2  40.80902 -73.94190  United States  ...       $124             3.0   
3  40.68514 -73.95976  United States  ...        $74            30.0   
4  40.79851 -73.94399  United States  ...        $41            10.0   

  number of reviews last review  reviews per month review rate number  \
0               9.0  10/19/2021               0.21                4.0   
1              45.0   5/21/2022               0.38                4.0   
2               0.0         NaN                NaN                5.0   
3             270.0    7/5/2019               4.64                4.0   
4               9.0  11/19/2018               0.10                3.0   

  calculated host listings count  availability 365  \
0                            6.0             286.0   
1                            2.0             228.0   
2                            1.0             352.0   
3                            1.0             322.0   
4                            1.0             289.0   

                                         house_rules license  
0  Clean up and treat the home the way you'd like...     NaN  
1  Pet friendly but please confirm with me if the...     NaN  
2  I encourage you to use my kitchen, cooking and...     NaN  
3                                                NaN     NaN  
4  Please no smoking in the house, porch or on th...     NaN  

[5 rows x 26 columns]

In [115]:
#information of data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102599 entries, 0 to 102598
Data columns (total 26 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   id                              102599 non-null  int64  
 1   NAME                            102349 non-null  object 
 2   host id                         102599 non-null  int64  
 3   host_identity_verified          102310 non-null  object 
 4   host name                       102193 non-null  object 
 5   neighbourhood group             102570 non-null  object 
 6   neighbourhood                   102583 non-null  object 
 7   lat                             102591 non-null  float64
 8   long                            102591 non-null  float64
 9   country                         102067 non-null  object 
 10  country code                    102468 non-null  object 
 11  instant_bookable                102494 non-null  object 
 12  cancellation_pol

In [116]:
#shape of data
print(f"the shape of this data is {data.shape}")

the shape of this data is (102599, 26)


In [117]:
#to show the variables
data.columns

Index(['id', 'NAME', 'host id', 'host_identity_verified', 'host name',
       'neighbourhood group', 'neighbourhood', 'lat', 'long', 'country',
       'country code', 'instant_bookable', 'cancellation_policy', 'room type',
       'Construction year', 'price', 'service fee', 'minimum nights',
       'number of reviews', 'last review', 'reviews per month',
       'review rate number', 'calculated host listings count',
       'availability 365', 'house_rules', 'license'],
      dtype='object')

DATA EXPLORATION

In [118]:
#Initial Percentage of Null Values Per Column
(data.isnull().sum()/len(data))*100

id                                 0.000000
NAME                               0.243667
host id                            0.000000
host_identity_verified             0.281679
host name                          0.395715
neighbourhood group                0.028265
neighbourhood                      0.015595
lat                                0.007797
long                               0.007797
country                            0.518524
country code                       0.127682
instant_bookable                   0.102340
cancellation_policy                0.074075
room type                          0.000000
Construction year                  0.208579
price                              0.240743
service fee                        0.266084
minimum nights                     0.398639
number of reviews                  0.178364
last review                       15.490404
reviews per month                 15.476759
review rate number                 0.317742
calculated host listings count  

In [119]:
#From the above cell where we checked for percentage of missing values in each column, any column with above 50% of missing observations will be dropped. 
#Therefore, house rules and license will be dropped since they have 51% & 99.99% NAN respectively. 
data.drop(columns=["house_rules", "license"], inplace=True)

In [120]:
#drop less important columns like ID, NAME 
data.drop(columns=["id","NAME"], inplace=True)

count the NAs for the remaining variables

In [121]:
data.isnull().sum()

host id                               0
host_identity_verified              289
host name                           406
neighbourhood group                  29
neighbourhood                        16
lat                                   8
long                                  8
country                             532
country code                        131
instant_bookable                    105
cancellation_policy                  76
room type                             0
Construction year                   214
price                               247
service fee                         273
minimum nights                      409
number of reviews                   183
last review                       15893
reviews per month                 15879
review rate number                  326
calculated host listings count      319
availability 365                    448
dtype: int64

FILLING THE NULL VALUES

i. The categorical columns will be filled through their modes 

ii. The integer columns will be filled using thier means


In [122]:
neighbourhood_gp_mode = data["neighbourhood group"].mode()
instant_b_mode = data["instant_bookable"].mode()
cancellation_gp = data["cancellation_policy"].mode()
room_mode = data["room type"].mode()
con_mode = data["Construction year"].mode()
hivc_mode = data["host_identity_verified"].mode()


print(f"the mode of neighbourhood group is {neighbourhood_gp_mode}")
print(f"the mode of instant bookable is {instant_b_mode}")
print(f"the mode of cancelltion is {cancellation_gp}")
print(f"the mode of room mode is {room_mode}")
print(f"the mode of construction year is {con_mode}")
print(f"the mode of HIVC is {hivc_mode}")

the mode of neighbourhood group is 0    Manhattan
Name: neighbourhood group, dtype: object
the mode of instant bookable is 0    False
Name: instant_bookable, dtype: object
the mode of cancelltion is 0    moderate
Name: cancellation_policy, dtype: object
the mode of room mode is 0    Entire home/apt
Name: room type, dtype: object
the mode of construction year is 0    2014.0
Name: Construction year, dtype: float64
the mode of HIVC is 0    unconfirmed
Name: host_identity_verified, dtype: object


In [123]:
#filling the null values of categorical variables with their modes. 
data["neighbourhood group"] = data["neighbourhood group"].fillna(value="Manhattan")
data["instant_bookable"] = data["instant_bookable"].fillna(value="False")
data["cancellation_policy"] = data["cancellation_policy"].fillna(value="moderate")
data["room type"] = data["room type"].fillna(value="Entire home/apt")
data["Construction year"] = data["Construction year"].fillna(value=2014.0)
data["host_identity_verified"] = data["host_identity_verified"].fillna(value="unconfirmed")


In [124]:
#Convert The Date to Month, Day and Year
data["last review"] = pd.to_datetime(data["last review"])
data["last review month"] = data["last review"].dt.month
data["last review day"] = data["last review"].dt.day
data["last review year"] = data["last review"].dt.year

data.head()

host id host_identity_verified host name neighbourhood group  \
0  80014485718            unconfirmed  Madaline            Brooklyn   
1  52335172823               verified     Jenna           Manhattan   
2  78829239556            unconfirmed     Elise           Manhattan   
3  85098326012            unconfirmed     Garry            Brooklyn   
4  92037596077               verified    Lyndon           Manhattan   

  neighbourhood       lat      long        country country code  \
0    Kensington  40.64749 -73.97237  United States           US   
1       Midtown  40.75362 -73.98377  United States           US   
2        Harlem  40.80902 -73.94190  United States           US   
3  Clinton Hill  40.68514 -73.95976  United States           US   
4   East Harlem  40.79851 -73.94399  United States           US   

  instant_bookable  ... minimum nights number of reviews  last review  \
0            False  ...           10.0               9.0   2021-10-19   
1            False  ...           30.0              45.0   2022-05-21   
2             True  ...            3.0               0.0          NaT   
3             True  ...           30.0             270.0   2019-07-05   
4            False  ...           10.0               9.0   2018-11-19   

  reviews per month review rate number  calculated host listings count  \
0              0.21                4.0                             6.0   
1              0.38                4.0                             2.0   
2               NaN                5.0                             1.0   
3              4.64                4.0                             1.0   
4              0.10                3.0                             1.0   

   availability 365 last review month  last review day  last review year  
0             286.0              10.0             19.0            2021.0  
1             228.0               5.0             21.0            2022.0  
2             352.0               NaN              NaN               NaN  
3             322.0               7.0              5.0            2019.0  
4             289.0              11.0             19.0            2018.0  

[5 rows x 25 columns]

In [125]:
#drop the last review column
data.drop(columns=["last review"], inplace=True)

In [126]:
data.isnull().sum()

host id                               0
host_identity_verified                0
host name                           406
neighbourhood group                   0
neighbourhood                        16
lat                                   8
long                                  8
country                             532
country code                        131
instant_bookable                      0
cancellation_policy                   0
room type                             0
Construction year                     0
price                               247
service fee                         273
minimum nights                      409
number of reviews                   183
reviews per month                 15879
review rate number                  326
calculated host listings count      319
availability 365                    448
last review month                 15893
last review day                   15893
last review year                  15893
dtype: int64

In [127]:
month_mode = data["last review month"].mode()
day_mode = data["last review day"].mode()
year_mode = data["last review year"].mode()

print(month_mode)
print(day_mode)
print(year_mode)

0    6.0
Name: last review month, dtype: float64
0    1.0
Name: last review day, dtype: float64
0    2019.0
Name: last review year, dtype: float64


In [128]:
data["last review month"] = data["last review month"].fillna(value=6.0)
data["last review day"] = data["last review day"].fillna(value=1.0)
data["last review year"] = data["last review year"].fillna(value=2019.0)

In [129]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102599 entries, 0 to 102598
Data columns (total 24 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   host id                         102599 non-null  int64  
 1   host_identity_verified          102599 non-null  object 
 2   host name                       102193 non-null  object 
 3   neighbourhood group             102599 non-null  object 
 4   neighbourhood                   102583 non-null  object 
 5   lat                             102591 non-null  float64
 6   long                            102591 non-null  float64
 7   country                         102067 non-null  object 
 8   country code                    102468 non-null  object 
 9   instant_bookable                102599 non-null  object 
 10  cancellation_policy             102599 non-null  object 
 11  room type                       102599 non-null  object 
 12  Construction yea

In [130]:
min_mean = np.round(data["minimum nights"].mean())
nor_mean = np.round(data["number of reviews"].mean())
rpm_mean = np.round(data["reviews per month"].mean())
rrn_mean = np.round(data["review rate number"].mean())
hol_mean = np.round(data["calculated host listings count"].mean())
av365_mean = np.round(data["availability 365"].mean())

data["minimum nights"] = data["minimum nights"].fillna(value=min_mean)
data["number of reviews"] = data["number of reviews"].fillna(value=nor_mean)
data["reviews per month"] = data["reviews per month"].fillna(value=rpm_mean)
data["calculated host listings count"] = data["calculated host listings count"].fillna(value=hol_mean)
data["availability 365"] = data["availability 365"].fillna(value=av365_mean)
data["review rate number"] = data["review rate number"].fillna(value=rrn_mean)

pr_mode = data["price"].mode()
sv_mode = data["service fee"].mode()

data["price"] = data["price"].fillna("$206")
data["service fee"] = data["service fee"].fillna("$41")

In [131]:
data.isnull().sum()

host id                             0
host_identity_verified              0
host name                         406
neighbourhood group                 0
neighbourhood                      16
lat                                 8
long                                8
country                           532
country code                      131
instant_bookable                    0
cancellation_policy                 0
room type                           0
Construction year                   0
price                               0
service fee                         0
minimum nights                      0
number of reviews                   0
reviews per month                   0
review rate number                  0
calculated host listings count      0
availability 365                    0
last review month                   0
last review day                     0
last review year                    0
dtype: int64

In [132]:
data["country"] = data["country"].fillna(value="United States")
data["country code"] = data["country code"].fillna(value="US")

In [133]:
data = data.dropna(axis=0)

In [134]:
data["last review month"].unique()

array([10.,  5.,  6.,  7., 11., 12.,  1.,  8.,  4.,  9.,  3.,  2.])

In [135]:
data["last review month"] = data["last review month"].map({
    10.:"October",  5.:"May",  6.:"June",  7.:"July", 11.:"November", 12.:"December",  1.:"January",  8.:"August",  4.:"April",  9.:"September",  3.:"March",  2.:"February"
})

In [136]:
data.head(11)

host id host_identity_verified host name neighbourhood group  \
0   80014485718            unconfirmed  Madaline            Brooklyn   
1   52335172823               verified     Jenna           Manhattan   
2   78829239556            unconfirmed     Elise           Manhattan   
3   85098326012            unconfirmed     Garry            Brooklyn   
4   92037596077               verified    Lyndon           Manhattan   
5   45498551794               verified  Michelle           Manhattan   
6   61300605564            unconfirmed   Alberta            Brooklyn   
7   90821839709            unconfirmed      Emma            Brooklyn   
8   79384379533               verified    Evelyn           Manhattan   
9   75527839483            unconfirmed      Carl           Manhattan   
10   1280143094               verified   Miranda           Manhattan   

         neighbourhood       lat      long        country country code  \
0           Kensington  40.64749 -73.97237  United States           US   
1              Midtown  40.75362 -73.98377  United States           US   
2               Harlem  40.80902 -73.94190  United States           US   
3         Clinton Hill  40.68514 -73.95976  United States           US   
4          East Harlem  40.79851 -73.94399  United States           US   
5          Murray Hill  40.74767 -73.97500  United States           US   
6   Bedford-Stuyvesant  40.68688 -73.95596  United States           US   
7   Bedford-Stuyvesant  40.68688 -73.95596  United States           US   
8       Hell's Kitchen  40.76489 -73.98493  United States           US   
9      Upper West Side  40.80178 -73.96723  United States           US   
10           Chinatown  40.71344 -73.99037  United States           US   

   instant_bookable  ... service fee minimum nights  number of reviews  \
0             False  ...       $193            10.0                9.0   
1             False  ...        $28            30.0               45.0   
2              True  ...       $124             3.0                0.0   
3              True  ...        $74            30.0              270.0   
4             False  ...        $41            10.0                9.0   
5              True  ...       $115             3.0               74.0   
6             False  ...        $14            45.0               49.0   
7             False  ...       $212            45.0               49.0   
8              True  ...       $204             2.0              430.0   
9             False  ...        $58             2.0              118.0   
10            False  ...        $64             1.0              160.0   

   reviews per month review rate number  calculated host listings count  \
0               0.21                4.0                             6.0   
1               0.38                4.0                             2.0   
2               1.00                5.0                             1.0   
3               4.64                4.0                             1.0   
4               0.10                3.0                             1.0   
5               0.59                3.0                             1.0   
6               0.40                5.0                             1.0   
7               0.40                5.0                             1.0   
8               3.47                3.0                             1.0   
9               0.99                5.0                             1.0   
10              1.33                3.0                             4.0   

    availability 365  last review month  last review day  last review year  
0              286.0            October             19.0            2021.0  
1              228.0                May             21.0            2022.0  
2              352.0               June              1.0            2019.0  
3              322.0               July              5.0            2019.0  
4              289.0           November             19.0            2018.0  
5 

In [137]:
#drop the lat, long amd neighbourhood column since they were used to get the neighborhood group
data.drop(columns=["neighbourhood", "lat", "long", "country"], inplace=True)


In [138]:
data.head()

host id host_identity_verified host name neighbourhood group  \
0  80014485718            unconfirmed  Madaline            Brooklyn   
1  52335172823               verified     Jenna           Manhattan   
2  78829239556            unconfirmed     Elise           Manhattan   
3  85098326012            unconfirmed     Garry            Brooklyn   
4  92037596077               verified    Lyndon           Manhattan   

  country code instant_bookable cancellation_policy        room type  \
0           US            False              strict     Private room   
1           US            False            moderate  Entire home/apt   
2           US             True            flexible     Private room   
3           US             True            moderate  Entire home/apt   
4           US            False            moderate  Entire home/apt   

   Construction year  price service fee  minimum nights  number of reviews  \
0             2020.0  $966        $193             10.0                9.0   
1             2007.0  $142         $28             30.0               45.0   
2             2005.0  $620        $124              3.0                0.0   
3             2005.0  $368         $74             30.0              270.0   
4             2009.0  $204         $41             10.0                9.0   

   reviews per month  review rate number  calculated host listings count  \
0               0.21                 4.0                             6.0   
1               0.38                 4.0                             2.0   
2               1.00                 5.0                             1.0   
3               4.64                 4.0                             1.0   
4               0.10                 3.0                             1.0   

   availability 365 last review month  last review day  last review year  
0             286.0           October             19.0            2021.0  
1             228.0               May             21.0            2022.0  
2             352.0              June              1.0            2019.0  
3             322.0              July              5.0            2019.0  
4             289.0          November             19.0            2018.0

In [139]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 102169 entries, 0 to 102598
Data columns (total 20 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   host id                         102169 non-null  int64  
 1   host_identity_verified          102169 non-null  object 
 2   host name                       102169 non-null  object 
 3   neighbourhood group             102169 non-null  object 
 4   country code                    102169 non-null  object 
 5   instant_bookable                102169 non-null  object 
 6   cancellation_policy             102169 non-null  object 
 7   room type                       102169 non-null  object 
 8   Construction year               102169 non-null  float64
 9   price                           102169 non-null  object 
 10  service fee                     102169 non-null  object 
 11  minimum nights                  102169 non-null  float64
 12  number of reviews    

FINAL CONCLUSION ON DATA CLEANING AND TRANSFORMATION

1. The columns with grater than 50% missing values were initially dropped
2. The columns which are not useful in the analysis were dropped
3. Three columns were created through the last review date column in order to sort out their missing values
4. The last review date year was transformed to strings (January, February.... December) for easy understanding
5. The columns which are categorical were filled with their most occuring values (MODE)
6. The columns which are integer were filled with their means.
7. The Host Name was as well used to drop some rows, as rows with no host names are considered unuseful for the analysis
8. The last dataset amount to a shape of 10269, 19. That is, 19 columns and 102169 rows. And can be used to generate more insights for Airbnb operations

POSSIBLE INSIGHTS TO GENRATE FROM THE NEW DATA

1. The neighbourhood group relationship with the host verification (Does neughborhood group affects hosts verification?)
2. The price and service fee based on neighborhood group
3. Relationship between review rates, neighbourhood group aqnd other services in the dataset
4. The host which has the best reviuew rates and the last time the host was reviewed, specifying their names and Ids